### DISCUSSION TITLE
get_history does not return attribute changes after Session.exectue()

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7603] [Usage Questions]
Hi

I try to update an object, using the `update` statement

```python
from sqlalchemy import update
from sqlalchemy import select
from sqlalchemy.orm.attributes import get_history

# select 
select_stmt = select(Table1).where(Table1.id == 1)
person1 = session.execute(select_stmt).scalar()
get_history(person1, 'name') # output History(added=(), unchanged=['Bob'], deleted=())
# update
updated_values = { 'name' : 'John' }
update_stmt = update(Table1).where(Table1.id == 1).values(**updated_values)
session.execute(update_stmt)

get_history(person1, 'name') # output History(added=(), unchanged=['John'], deleted=())
``` 
I expected the second history to show me somehow that the 'name' property was changed from 'Bob' to 'John' but after the execute statement, even though the object has been updated, the history seems to be a pretty new one (since 'John', which is the new value for  'name' is mentioned as 'unchanged')

What did I miss here?

**Comment by zzzeek** [comment]
the history collection for attributes is a short term, transitory thing that only lasts until the next time the session flushes.  the session then autoflushes when the .execute() method is called, when an ORM-enabled statement is passed to it.

if you want to grab history for objects before it goes away the [before_flush](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.before_flush) event is one way this could be done.   you can also intercept changes on individual attributes using [AttributeEvents](https://docs.sqlalchemy.org/en/14/orm/events.html#attribute-events).  there's also ways to set up attribute listeners on all attributes for a class / hierarchy, such as the example at [listen_for_events](https://docs.sqlalchemy.org/en/14/_modules/examples/custom_attributes/listen_for_events.html).
